ocaml/utils/global.ml

```diff
@@ -0,0 +1,146 @@
+[@@@ocaml.warning "+a-40-41-42"]
+
+let pp_concat pp ppf list =
+  Format.pp_print_list ~pp_sep:Format.pp_print_cut pp ppf list
+
+module Name = struct
+  type t = {
+    head : string;
+    args : (t * t) list;
+  }
+
+  include Identifiable.Make (struct
+    type nonrec t = t
+
+    let rec compare
+        { head = head1; args = args1 }
+        { head = head2; args = args2 } =
+      match String.compare head1 head2 with
+      | 0 -> List.compare compare_arg args1 args2
+      | c -> c
+    and compare_arg (name1, arg1) (name2, arg2) =
+      match compare name1 name2 with
+      | 0 -> compare arg1 arg2
+      | c -> c
+
+    let equal t1 t2 = t1 == t2 || compare t1 t2 = 0
+
+    let rec print ppf ({ head; args } : t) =
+      Format.fprintf ppf "@[<hov 1>%s%a@]"
+        head
+        (pp_concat print_arg_pair) args
+    and print_arg_pair ppf (name, arg) =
+      Format.fprintf ppf "[%a:%a]" print name print arg
+
+    let output = print |> Misc.output_of_print
+
+    let rec hash { head; args } =
+      Hashtbl.hash (head, List.map hash_arg args)
+    and hash_arg (name, param) =
+      Hashtbl.hash (hash name, hash param)
```

Comment:
I think this is just the same as defining `let hash = Hashtbl.hash`.

Reply:
Changed.